js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/useShippingRatesWithSavedSuggestions.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * External dependencies
+ */
+import { useState, useRef, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useShippingRates from '.~/hooks/useShippingRates';
+import useShippingRatesSuggestions from './useShippingRatesSuggestions';
+import useSaveSuggestions from './useSaveSuggestions';
+import useCallbackOnceEffect from '.~/hooks/useCallbackOnceEffect';
+
+/**
+ * @typedef {Object} ShippingRatesWithSavedSuggestionsResult
+ * @property {boolean} loading Whether loading is in progress.
+ * @property {Array<import('.~/data/actions').ShippingRate>?} data Shipping rates.
+ */
+
+/**
+ * Check existing shipping rates, and if it is empty, get shipping rates suggestions
+ * and save the suggestions as shipping rates.
+ *
+ * @return {ShippingRatesWithSavedSuggestionsResult} Result object with `loading` and `data`.
+ */
+const useShippingRatesWithSavedSuggestions = () => {
+	const {
+		hasFinishedResolution: hfrShippingRates,
+		data: dataShippingRates,
+	} = useShippingRates();
+	const {
+		loading: loadingSuggestions,
+		data: dataSuggestions,
+	} = useShippingRatesSuggestions();
+
+	/**
+	 * `isInitialShippingRatesEmpty` is used to indicate
+	 * whether the initial loaded shipping rates
+	 * has a pre-saved value or not.
+	 *
+	 * If it does not have a pre-saved value,
+	 * shipping rates should be an empty array,
+	 * and we should save the suggestions as shipping rates.
+	 *
+	 * If it does have a pre-saved value,
+	 * then we should not save the suggestions,
+	 * even when users manually deleted all the pre-saved shipping rates value.
+	 * The exception is when users deleted all the pre-saved value
+	 * and then reload the page,
+	 * then the suggestions would be saved as shipping rates as per above logic.
+	 */
+	const isInitialShippingRatesEmpty = useRef( undefined );
+	if (
+		hfrShippingRates &&
+		isInitialShippingRatesEmpty.current === undefined
+	) {
+		isInitialShippingRatesEmpty.current = dataShippingRates.length === 0;
+	}
+
+	/**
+	 * `done` is used to indicate whether saving has finished.
+	 * This is only used when we have no pre-saved initial shipping rates value
+	 * and we call `saveSuggestions`. It is initially set to `false`,
+	 * and will be set to `true` after the suggestions are saved.
+	 */
+	const [ done, setDone ] = useState( false );
+	const saveSuggestions = useSaveSuggestions();
+	const callSaveSuggestions = useCallback(
+		async ( suggestions ) => {
+			await saveSuggestions( suggestions );
+			setDone( true );
+		},
+		[ saveSuggestions ]
+	);
+
+	/**
+	 * `callSaveSuggestions` when:
+	 *
+	 * - there is no pre-saved initial shipping rates value.
+	 * - we have suggestions data.
+	 */
+	useCallbackOnceEffect(
+		isInitialShippingRatesEmpty.current && dataSuggestions,
+		callSaveSuggestions,
+		dataSuggestions
+	);
+
+	return {
+		loading:
+			loadingSuggestions ||
+			! hfrShippingRates ||
+			( isInitialShippingRatesEmpty.current ? ! done : false ),
```

Comment:
❓ 💅 

As far as I understood the code. 

`loading: ! done || ! hfrShippingRates` would be sufficient ?? . 

- `done` only can be true if `isInitialShippingRatesEmpty.current` is true, because is a condition for the new hook `useCallbackOnceEffect` 

- `done` only can be true if `dataSuggestions` is populated as it's a condition for `useCallbackOnceEffect`,  which means `loadingSuggestions` is done? Not sure about this one tho. 

If, we are calling `useCallbackOnceEffect` after `dataSuggestions` is populated but not necessary means loaded is completed... does it mean we can call at some point `useCallbackOnceEffect` with uncompleted data??? Not sure about this one tbh.

I understood that even after having the first `hfrShippingRatescould` completed, we are updating them potentially with `saveSuggestions` hence, `hfrShippingRatescould` could be updated.  Thats why it's required in this condition.

Reply:
Nevermind, after checking again the tests and the code ... it seems like `isInitialShippingRatesEmpty.current` is as well necessary. because we don't care about `done` in some cases 

Since it was confusing to read, I would suggest 
`(isInitialShippingRatesEmpty.current && ! done)` 

or even better 

`(shouldSaveSuggestions && ! suggestionsSaved)`

Reply:
Simplified the conditional code in [`ff3b876` (#1161)](https://github.com/woocommerce/google-listings-and-ads/pull/1161/commits/ff3b87662963333b1044a7f90099fdf3cb50e10d).